Make magnetometer calibration independent of locale (Fixes #452)

thp opened the pull request:
Force the `LC_NUMERIC` locale to be the `C` locale, so that `%f` writes/reads `.` always for floating point (and not `,`). As part of this change, also cleaned up the actual read/write code, as it was _much_ more complicated than necessary.

(Historical tidbit: The original code was written by me nearly 10 years ago, see bde8a6a20431261400a2f55f65243e444c17ca76 - quite silly in retrospect to do it like that when `printf()` and `scanf()` are perfectly capable of writing/reading verbatim characters...)

Not yet handled: Unsure how Win32 handles locales, it definitely doesn't seem to have `uselocale()` / `newlocale()` / `freelocale()` from a quick glance at mingw-w64 headers.

thp commented:
For Win32, something like this could work, but I wasn't able to test it, as Wine doesn't implement `_printf_l` in its MSVCRT:

```c
#include <stdio.h>
#include <locale.h>

int main()
{
    setlocale(LC_NUMERIC, "de_AT");

    printf("Hello, world! %f\n", 12.34f);

    _locale_t loc = _create_locale(LC_NUMERIC, "C");
    _printf_l("Hello, world! %f\n", loc, 12.34f);
    _free_locale(loc);

    return 0;
}
```

(build with `i686-w64-mingw32-gcc -o test test.c`)

This fails with:

```
% wine test.exe
Hello, world! 12,340000
wine: Call from 7BC2CEA0 to unimplemented function msvcrt.dll._printf_l, aborting
```

nitsch commented:
> For Win32, something like this could work, but I wasn't able to test it, as Wine doesn't implement `_printf_l` in its MSVCRT:

Yeah, it does. It has to be `de-AT` instead of `de_AT` to set the "wrong" locale, but the relevant part seems to work on Windows.

thp commented:
> Yeah, it does. It has to be `de-AT` instead of `de_AT` to set the "wrong" locale, but the relevant part seems to work on Windows.

Added the code. Let's see if it builds in CI.

nitsch commented on src/psmove.c:
```diff
@@ -45,6 +45,15 @@
 #include <limits.h>
+#include <stdarg.h>
+
+#if defined(__linux) || defined(__APPLE__)
```
Shouldn't that just check for "not Apple" to include locale.h for Linux and Win32? Apple is covered in the `#ifdef` below.

thp replied:
Yep, and that actually broke Win32 builds, ALL platforms need `locale.h`, and macOS also needs `xlocale.h`. Fixed now.

nitsch commented on src/psmove.c:
```diff
@@ -303,6 +312,44 @@ static int psmove_remote_disabled = 0;
+    va_end(args);
+    return result;
+}
+
```
No error checking here. I don't love that, but it might not be that big of a deal.

thp replied:
Error checking for `psmove_fscanf_c()` is done by checking the return value against `PSMOVE_MAGNETOMETER_CALIBRATION_FSTRING_FIELDS` on the call site.

Error checking for `psmove_fprintf_c()` isn't done at the moment (as it returns the number of bytes written, not how many fields have been written).

thp replied:
(or did you mean error checking for `_create_locale`, `_free_locale`, `uselocale`, `newlocale` and `freelocale`?)

nitsch replied:
Yes, that's what I meant.

thp replied:
Added error checking for `newlocale()` now. The other functions seem to not be able to fail. Also, the Win32 APIs seem to only be available when linking against the Universal C Runtime (UCRT), so special-cased this, with a fallback (e.g. mingw-w64) to just the normal vfprintf/vfscanf functions. Eventually we should look into building against the UCRT on Windows. However, it could be that MSVCRT doesn't even have issues with locales?

nitsch replied:
On  the off-chance that the whole thing is unnecessary with the MSVCRT, I will look into this later today.

Not sure about the UCRT issue, though. Doesn't that just complain about `_vfscanf_l` not being available? `_fscanf_l` should be fine. I mean, we could always just do all the platform-dependent special handling for that one particular format string we are dealing with and not try to generalize it.

thp replied:
Interesting that `_fscanf_l` would be available, but that seems to be what the mingw headers tell me too.. weird.

Tried to take that into account in d37b68d523bf20b53b6da49e18dd9a381063357e (but untested, as I don't have a Windows build machine around, but CI should catch it).

nitsch commented on src/psmove.c:
```diff
@@ -303,6 +310,66 @@ static int psmove_remote_disabled = 0;
+    _locale_t loc = _create_locale(LC_NUMERIC, "C");
+    result = _vfscanf_l(fp, fmt, loc, args);
+    _free_locale(loc);
+#elif defined(_WIN32)
+    result = vfscanf(fp, fmt, args);
+#else
```
Unsurprisingly, MSVCRT also needs the locale treatment. So this special case will result in the program misbehaving on those platforms.

thp replied:
This is taken care of in d37b68d523bf20b53b6da49e18dd9a381063357e -- if it works :)

thp commented:
At this point, I'm quite tempted to just put the data into a struct, dump that struct to the calibration file and read the struct back from the calibration file as binary data (ignoring alignment and endianness issues, as it should be read back on the same machine, anyway, and between 32-bit and 64-bit x86 as well as 32-bit and 64-bit ARM, alignment and endianness are well-defined and should match up well enough for this to not be another pain point), as the OS-specific handling of this is ugly, and especially Win32 is the odd one out. Also, switching to a binary format without having backwards compatible code means people need to re-calibrate their controllers when upgrading the library.

I tried to avoid repetition by using macros, but it doesn't really improve readability.

If it builds like this, might merge it as-is.

nitsch commented:
> At this point, I'm quite tempted to just put the data into a struct, dump that struct to the calibration file and read the struct back from the calibration file as binary data (ignoring alignment and endianness issues, as it should be read back on the same machine, anyway, and between 32-bit and 64-bit x86 as well as 32-bit and 64-bit ARM, alignment and endianness are well-defined and should match up well enough for this to not be another pain point)

Yeah, I also considered this option when it became apparent that the whole special-casing was necessary to do it the right way [TM]. Storing the calibration data in human readable format is not actually of much use since nobody is going to tweak these values in a meaningful way anyway, I suppose. I would have tried to take care of the endianness, though (either by defining a fixed byte order for the file an parsing accordingly or by dumping a known word along with the struct as indicator). But you have a point there.

thp commented:
> > At this point, I'm quite tempted to just put the data into a struct, dump that struct to the calibration file and read the struct back from the calibration file as binary data (ignoring alignment and endianness issues, as it should be read back on the same machine, anyway, and between 32-bit and 64-bit x86 as well as 32-bit and 64-bit ARM, alignment and endianness are well-defined and should match up well enough for this to not be another pain point)
> 
> Yeah, I also considered this option when it became apparent that the whole special-casing was necessary to do it the right way [TM]. Storing the calibration data in human readable format is not actually of much use since nobody is going to tweak these values in a meaningful way anyway, I suppose. I would have tried to take care of the endianness, though (either by defining a fixed byte order for the file and parsing accordingly or by dumping a known word along with the struct as indicator). But you have a point there.

Haha, we nearly got there, but in the end the mingw32 build failed, and while it'd probably be fine to drop 32-bit Windows support at this point, let's just go for the binary format instead. I've added a magic number at the start of the file, so in the future (not yet implemented) if we ever wanted to support reading calibration files from different endianness, we could add support there (the magic number's byte order would tell us whether or not we would need to byte-swap on read).

thp commented:
Squashed some commits and fixed the comment. The locale-related commits I'll leave in the history if we never need to dig it out. Let's see if it builds, and would be good of somebody could test-drive the new code (don't have a Move setup here at the moment).

thp commented:
CI seems to be happy now. @nitsch if you have time for a quick review/approval, please do (no rush). I'll merge after you're happy with the changes.

thp commented:
Merged this after Alexander has closed this as completed: https://github.com/thp/psmoveapi/issues/452#event-7648313934
